src/trampoline.c

```diff
@@ -73,7 +77,12 @@ plt_fn *create_trampoline(void *ctx, plt_fn *routine)
     *map = ctx;
     *(map + 1) = (void *) routine;
     memcpy(map + 2, mmk_trampoline, trampoline_sz);
-    mprotect(map, PAGE_SIZE, PROT_READ | PROT_EXEC);
+    mmk_assert(!mprotect(map, PAGE_SIZE, PROT_READ | PROT_EXEC));
+# if defined __clang__  // Check for Clang first, it may set __GNUC__ too.
+    __clear_cache(map, map + PAGE_SIZE);
+# elif defined __GNUC__
+    __builtin___clear_cache((char *)map, (char *)(map + PAGE_SIZE));
+# endif
```

Comment:
Interesting. Do you know why clearing the cache is necessary?

The mprotect in `plt_set_offsets` didn't seem to crash, is a cache clear not necessary in that function? This is weird.

Reply:
> Interesting. Do you know why clearing the cache is necessary?
>
> The mprotect in `plt_set_offsets` didn't seem to crash, is a cache clear not necessary in that function? This is weird.

So, not clearing caches does not consistently make the library crash. I have not had issues with Mimick as-is on `amd64`. But on `aarch64` I'd get a SIGILL somewhere within the trampoline ([here](https://github.com/Snaipe/Mimick/blob/7681d1ee8e0bffff806751092947648a6bebff24/src/asm/trampoline-aarch64.S#L27) and [here](https://github.com/Snaipe/Mimick/blob/7681d1ee8e0bffff806751092947648a6bebff24/src/asm/trampoline-aarch64.S#L68), for instance). I was initially suspicious about `mprotect`. Eventually I found out that [`glibc` does flush caches](https://github.com/bminor/glibc/blob/master/elf/dl-reloc.c#L294) when bringing in a shared library code into memory space. I can't tell for sure, but I suspect it's the instruction cache that's causing problems when flipping page protection bits.

> The mprotect in plt_set_offsets didn't seem to crash, is a cache clear not necessary in that function?

Uh, I missed that one. Also, `mmk_mprotect` should be used here and not `mprotect`. See https://github.com/Snaipe/Mimick/pull/10/commits/edd0150e328c2630f0ac5eef0f546ff1ec291a07.